src/daft-table/src/ops/joins/mod.rs

```diff
@@ -121,6 +122,166 @@ pub fn infer_join_schema(
+                .collect::<Vec<_>>(),
+        )?
+        .into();
+        let right_schema = Schema::new(
+            mapping
+                .iter()
+                .filter(|jc| jc.is_right)
+                .map(|jc| jc.field.clone())
+                .collect::<Vec<_>>(),
+        )?
+        .into();
+        Ok(JoinOutputMapper {
+            mapping,
+            left_schema,
+            right_schema,
+        })
+    }
+
+    pub fn map_left(&self, table: &Table) -> DaftResult<Table> {
+        let out = self
+            .mapping
+            .iter()
+            .filter(|jc| !jc.is_right)
+            .map(|jc| DaftResult::Ok(table.get_column_by_index(jc.index)?.rename(&jc.field.name)))
+            .collect::<DaftResult<Vec<_>>>()?;
+
+        Table::new_with_size(self.left_schema.clone(), out, table.num_rows)
+    }
+
+    pub fn map_right(&self, table: &Table) -> DaftResult<Table> {
+        let out = self
+            .mapping
+            .iter()
+            .filter(|jc| jc.is_right)
+            .map(|jc| DaftResult::Ok(table.get_column_by_index(jc.index)?.rename(&jc.field.name)))
+            .collect::<DaftResult<Vec<_>>>()?;
+
+        Table::new_with_size(self.right_schema.clone(), out, table.num_rows)
+    }
+}
+
+pub fn infer_join_schema_mapper(
+    left: &SchemaRef,
+    right: &SchemaRef,
+    left_on: &[ExprRef],
+    right_on: &[ExprRef],
+    how: JoinType,
+) -> DaftResult<JoinOutputMapper> {
+    if left_on.len() != right_on.len() {
+        return Err(DaftError::ValueError(format!(
+            "Length of left_on does not match length of right_on for Join {} vs {}",
+            left_on.len(),
+            right_on.len()
+        )));
+    }
+    if matches!(how, JoinType::Anti | JoinType::Semi) {
+        return JoinOutputMapper::try_new(
+            left.fields
+                .values()
+                .into_iter()
+                .enumerate()
+                .map(|(i, f)| JoinOutputColumn {
+                    is_right: false,
+                    index: i,
+                    field: f.clone(),
+                })
+                .collect(),
+        );
+    }
+    let mut result = vec![];
+    let mut names_so_far = HashSet::new();
+
+    let lfields = left_on
+        .iter()
+        .map(|e| e.to_field(left))
+        .collect::<DaftResult<Vec<_>>>()?;
+    let rfields = right_on
+        .iter()
+        .map(|e| e.to_field(right))
+        .collect::<DaftResult<Vec<_>>>()?;
+
+    for lf in lfields.iter() {
+        let name = &lf.name;
+        let index = left.get_index(name)?;
+        let field = left.get_field(name)?.clone();
+        result.push(JoinOutputColumn {
+            is_right: false,
+            index,
+            field,
+        });
+        names_so_far.insert(name.clone());
+    }
+
+    let left_names = lfields.iter().map(|e| e.name.as_str());
+    let right_names = rfields.iter().map(|e| e.name.as_str());
+
+    // Then Add Left Table non-join-key columns
```

Comment:
So it's a bit more complicated than that because someone may have a join like:
```
df.join(df2, left_on = col("x") * 2, right_on=col("y").cast(int))
```
in this case, we want to use the fields (and types)) from the original schema but emit out the join key from both sides due to the difference in names.